src/Microsoft.VisualStudio.Threading/AsyncReaderWriterResourceLock`2.cs

```diff
@@ -760,14 +761,35 @@ private Task PrepareResourceAsync(TResource resource, CancellationToken cancella
                             finalState);
                     }
                 }
+                else
+                {
+                    joinToExistingTask = true;
+                }
 
                 Assumes.NotNull(preparationTask.PreparationTask);
                 this.resourcePreparationTasks[resource] = preparationTask;
 
+                Task computationTask = preparationTask.PreparationTask;
+
+                if (forConcurrentUse && joinToExistingTask)
+                {
+                    return computationTask.ContinueWith(
+                        t =>
+                        {
+                            if (t.IsFaulted)
```

Comment:
What if the antecedent is canceled? Do you need to propagate that as well?

Reply:
make sense, I fixed it.